pageserver/src/tenant/layer_map.rs

```diff
@@ -499,87 +263,237 @@ where
-            let img = self.find_latest_image(start, lsn);
-
-            ranges.push((start..*end, img));
-
-            start = *end;
-        }
-        Ok(ranges)
+        Ok(coverage)
     }
 
-    /// Count how many L1 delta layers there are that overlap with the
-    /// given key and LSN range.
-    pub fn count_deltas(&self, key_range: &Range<Key>, lsn_range: &Range<Lsn>) -> Result<usize> {
-        let mut result = 0;
-        if lsn_range.start >= lsn_range.end {
+    /// Count the height of the tallest stack of deltas in this 2d region.
+    ///
+    /// If `limit` is provided we don't try to count above that number.
+    ///
+    /// This number is used to compute the largest number of deltas that
+    /// we'll need to visit for any page reconstruction in this region.
+    /// We use this heuristic to decide whether to create an image layer.
+    pub fn count_deltas(
+        &self,
+        key: &Range<Key>,
+        lsn: &Range<Lsn>,
+        limit: Option<usize>,
+    ) -> Result<usize> {
+        // We get the delta coverage of the region, and for each part of the coverage
+        // we recurse right underneath the delta. The recursion depth is limited by
+        // the largest result this function could return, which is in practice between
+        // 3 and 10 (since we usually try to create an image when the number gets larger).
+
+        if lsn.is_empty() || key.is_empty() || limit == Some(0) {
             return Ok(0);
         }
-        let envelope = AABB::from_corners(
-            [
-                IntKey::from(key_range.start.to_i128()),
-                IntKey::from(lsn_range.start.0 as i128),
-            ],
-            [
-                IntKey::from(key_range.end.to_i128() - 1),
-                IntKey::from(lsn_range.end.0 as i128 - 1),
-            ],
-        );
-        for e in self
-            .historic_layers
-            .locate_in_envelope_intersecting(&envelope)
-        {
-            let l = &e.layer;
-            if !l.is_incremental() {
-                continue;
+
+        let version = match self.historic.get().unwrap().get_version(lsn.end.0 - 1) {
+            Some(v) => v,
+            None => return Ok(0),
+        };
+
+        let start = key.start.to_i128();
+        let end = key.end.to_i128();
+
+        // Initialize loop variables
+        let mut max_stacked_deltas = 0;
+        let mut current_key = start;
+        let mut current_val = version.delta_coverage.query(start);
+
+        // Loop through the delta coverage and recurse on each part
+        for (change_key, change_val) in version.delta_coverage.range(start..end) {
+            // If there's a relevant delta in this part, add 1 and recurse down
+            if let Some(val) = current_val {
+                if val.get_lsn_range().end > lsn.start {
+                    let kr = Key::from_i128(current_key)..Key::from_i128(change_key);
+                    let lr = lsn.start..val.get_lsn_range().start;
```

Comment:
One small notice: `get_lsn_range/get_key_range` are not so cheap functions as may be expected.
Them perform cloning of range and I quite often saw them in profile.
I though how it wil be possible to avoid clone here by returning reference, but InMemoryLayer constructs range dynamically so it is not possible.
At least it willbe good to assign it to local variable to avoid redudant calls.

Reply:
> but InMemoryLayer constructs range dynamically so it is not possible

It's still possible. Construct it dynamically and return a reference to that, just to comply with the API.

Reply:
It seems to me that it is not possible (returning dynamically allocated value as reference) because it is obvious memory leak. We can use boxing, but I do not like it.

Reply:
Hm I thought that the lifetime of the local variables would contain the lifetime of the returned value. But it's not how it works (maybe for a good reason?) It only works if you directly construct the end result in one step.

I guess `Cow` is another option.

Or if we really want 0-overhead, I bet it's possible after https://github.com/neondatabase/neon/pull/3263/files (or similar) refactor. `InMemoryLayer` has no business being in the layer map, so it doesn't need to implement `LayerRange`, as defined in that PR.

Anyway, i'd leave this as a follow-up to this PR since even with all that cloning this PR is not a perf degradation.